horovod/spark/common/cache.py

```diff
@@ -0,0 +1,80 @@
+# Copyright 2019 Uber Technologies, Inc. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+
+from __future__ import absolute_import
+
+import threading
+from contextlib import contextmanager
+
+
+class TrainingDataCache(object):
+    def __init__(self):
+        self._lock = threading.Lock()
+        self._reset()
+
+    @contextmanager
+    def acquire_lock_for_key(self, key):
+        with self._lock:
+            if key not in self._entries:
+                self._entry_locks[key] = threading.Lock()
+                self._entries[key] = None
+
+        lock = self._entry_locks[key]
+        lock.acquire()
+        try:
+            yield
+        finally:
+            lock.release()
+
+    def get(self, key):
+        return self._entries[key]
+
+    def put(self, key, value):
+        self._entries[key] = value
+
+    def is_cached(self, key):
+        dataframe_hash, validation_split, validation_col, train_data_path, val_data_path = key
+
+        content = self._entries.get(key, None)
+        if not content:
+            return False
+
+        return \
+            content.dataframe_hash == dataframe_hash and \
+            content.validation_split == validation_split and \
+            content.validation_col == validation_col and \
+            content.store.exists(train_data_path) and \
+            (validation_split == 0.0 or content.store.exists(val_data_path))
+
+    def clear(self):
+        self._reset()
+
+    def _reset(self):
+        with self._lock:
+            self._entry_locks = dict()
```

Comment:
race condition: if while one thread calls into `acquire_lock_for_key` another thread calls `clear`, the `self._lock` is only held over lines 30-32, setting `self._entry_locks = dict` can happen right before `self._entry_locks[key]`, which will then throw an exception. That exception might not be very descriptive (acquiring the lock failed, but the exceptions says something about a missing key in a dictionary), if desired at all.

Reply:
Good catch!  I think having a per-key lock in this instance is overkill.  Removed and replaced with a single global lock.